File: nodejs/fibonacci.js
#!/usr/bin/env node

/**
 * Enhanced Fibonacci Sequence Generator in Node.js
 */

const readline = require('readline');

// ANSI color codes
const RESET = '\x1b[0m';
const GREEN = '\x1b[32m';
const YELLOW = '\x1b[33m';
const CYAN = '\x1b[36m';
const MAGENTA = '\x1b[35m';
const RED = '\x1b[31m';
const GRAY = '\x1b[90m';

// Global variables
let t1, t2, nextTerm, sum;
let maxValue;
let ratio;

/**
 * Clear the terminal screen
 */
function clearScreen() {
    // Use ANSI escape codes to clear screen
    console.clear();
}

/**
 * Display the program header
 */
function displayHeader() {
    console.log(`${GREEN}===============================================${RESET}`);
    console.log(`${GREEN}      Enhanced Fibonacci Sequence Generator    ${RESET}`);
    console.log(`${GREEN}===============================================${RESET}`);
    console.log();
}

/**
 * Display menu options
 */
function displayMenu() {
    console.log(`${YELLOW}Generation Options:${RESET}`);
    console.log('  1. Generate by number of terms');
    console.log('  2. Generate up to a maximum value');
    console.log('  3. Display with statistics');
    console.log('  4. Display in columns');
    console.log();
    process.stdout.write('Enter your choice (1-4): ');
}

/**
 * Generate and display Fibonacci sequence
 * @param {number} termCount - Number of terms to generate
 * @param {boolean} showStats - Whether to show statistics
 * @param {boolean} columnMode - Whether to display in columns
 */
function generateFibonacci(termCount, showStats, columnMode) {
    t1 = 0n;
    t2 = 1n;
    sum = 0n;
    const termsPerLine = 10;
    let output = '';

    if (columnMode) {
        console.log(`${CYAN}Fibonacci Sequence (${termCount} terms):${RESET}`);
        console.log();
    } else {
        console.log(`${CYAN}Fibonacci Sequence:${RESET}`);
    }

    // Handle the first term
    if (termCount >= 1) {
        if (columnMode) {
            output += t1.toString().padStart(12);
        } else {
            output += `${t1} `;
        }
        sum += t1;
    }

    // Handle the second term
    if (termCount >= 2) {
        if (columnMode) {
            output += t2.toString().padStart(12);
        } else {
            output += `${t2} `;
        }
        sum += t2;
    }

    // Calculate and print remaining terms
    for (let j = 3; j <= termCount; j++) {
        nextTerm = t1 + t2;

        if (columnMode) {
            output += nextTerm.toString().padStart(12);
            if (j % termsPerLine === 0) {
                output += '\n';
            }
        } else {
            output += `${nextTerm} `;
        }

        sum += nextTerm;

        t1 = t2;
        t2 = nextTerm;
    }

    console.log(output);
    console.log();

    // Display statistics if requested
    if (showStats) {
        console.log(`${MAGENTA}Statistics:${RESET}`);
        console.log(`  Number of terms: ${termCount}`);
        console.log(`  Sum of all terms: ${sum}`);
        const average = Number(sum) / termCount;
        console.log(`  Average value: ${average.toFixed(2)}`);

        if (termCount >= 2) {
            console.log(`  Last term: ${t2}`);
            console.log(`  Second-to-last term: ${t1}`);

            if (t1 !== 0n) {
                ratio = Number(t2) / Number(t1);
                const goldenRatio = 1.6180339887;
                const difference = Math.abs(ratio - goldenRatio);
                console.log(`  Ratio (F(n)/F(n-1)): ${ratio.toFixed(10)}`);
                console.log(`  Golden Ratio (φ): ${goldenRatio.toFixed(10)}`);
                console.log(`  Difference: ${difference.toFixed(10)}`);
            }
        }
        console.log();
    }
}

/**
 * Generate Fibonacci sequence up to a maximum value
 * @param {bigint} maxVal - Maximum value to generate up to
 */
function generateUpToMax(maxVal) {
    t1 = 0n;
    t2 = 1n;
    let count = 0;
    let output = '';

    console.log(`${CYAN}Fibonacci Sequence up to ${maxVal}:${RESET}`);
    console.log();

    if (maxVal >= 0n) {
        output += `${t1} `;
        count++;
    }

    if (maxVal >= 1n) {
        output += `${t2} `;
        count++;
    }

    while ((t1 + t2) <= maxVal) {
        nextTerm = t1 + t2;
        output += `${nextTerm} `;
        count++;
        t1 = t2;
        t2 = nextTerm;
    }

    console.log(output);
    console.log();
    console.log(`${MAGENTA}Total terms generated: ${count}${RESET}`);
    console.log();
}

/**
 * Parse integer input safely
 * @param {string} input - Input string to parse
 * @returns {number|null} Parsed number or null if invalid
 */
function parseIntSafe(input) {
    const num = parseInt(input.trim(), 10);
    return isNaN(num) ? null : num;
}

/**
 * Parse bigint input safely
 * @param {string} input - Input string to parse
 * @returns {bigint|null} Parsed bigint or null if invalid
 */
function parseBigIntSafe(input) {
    try {
        return BigInt(input.trim());
    } catch (e) {
        return null;
    }
}

/**
 * Main program entry point
 */
async function main() {
    const rl = readline.createInterface({
        input: process.stdin,
        output: process.stdout
    });

    // Promisify readline question
    const question = (query) => new Promise((resolve) => rl.question(query, resolve));

    clearScreen();
    displayHeader();
    displayMenu();

    const choiceInput = await question('');
    const choice = parseIntSafe(choiceInput);
    console.log();

    try {
        switch (choice) {
            case 1: {
                const nInput = await question('Enter the number of terms to generate (max 1476): ');
                let n = parseIntSafe(nInput);
                console.log();

                if (n === null || n <= 0) {
                    console.log(`${RED}Error: Please enter a positive integer.${RESET}`);
                } else if (n > 1476) {
                    console.log(`${RED}Error: Maximum 1476 terms supported with BigInt.${RESET}`);
                    console.log('Using 1476 terms instead.');
                    n = 1476;
                    console.log();
                    generateFibonacci(n, false, false);
                } else {
                    generateFibonacci(n, false, false);
                }
                break;
            }

            case 2: {
                const maxInput = await question('Enter the maximum value: ');
                const maxValue = parseBigIntSafe(maxInput);
                console.log();

                if (maxValue === null || maxValue < 0n) {
                    console.log(`${RED}Error: Please enter a non-negative value.${RESET}`);
                } else {
                    generateUpToMax(maxValue);
                }
                break;
            }

            case 3: {
                const nInput = await question('Enter the number of terms to generate (max 1476): ');
                let n = parseIntSafe(nInput);
                console.log();

                if (n === null || n <= 0) {
                    console.log(`${RED}Error: Please enter a positive integer.${RESET}`);
                } else if (n > 1476) {
                    console.log(`${RED}Error: Maximum 1476 terms supported with BigInt.${RESET}`);
                    console.log('Using 1476 terms instead.');
                    n = 1476;
                    console.log();
                    generateFibonacci(n, true, false);
                } else {
                    generateFibonacci(n, true, false);
                }
                break;
            }

            case 4: {
                const nInput = await question('Enter the number of terms to generate (max 1476): ');
                let n = parseIntSafe(nInput);
                console.log();

                if (n === null || n <= 0) {
                    console.log(`${RED}Error: Please enter a positive integer.${RESET}`);
                } else if (n > 1476) {
                    console.log(`${RED}Error: Maximum 1476 terms supported with BigInt.${RESET}`);
                    console.log('Using 1476 terms instead.');
                    n = 1476;
                    console.log();
                    generateFibonacci(n, false, true);
                } else {
                    generateFibonacci(n, false, true);
                }
                break;
            }

            default: {
                console.log(`${RED}Invalid choice. Using default mode (option 1).${RESET}`);
                console.log();
                const nInput = await question('Enter the number of terms to generate (max 1476): ');
                let n = parseIntSafe(nInput);
                console.log();

                if (n !== null && n > 0) {
                    if (n > 1476) n = 1476;
                    generateFibonacci(n, false, false);
                }
                break;
            }
        }

        console.log();
        console.log(`${GRAY}Press ENTER to exit...${RESET}`);
        await question('');
    } finally {
        rl.close();
    }
}

// Run the main program
main().catch(console.error);

File: rust/fibonacci.rs
#!/usr/bin/env rust-script

/**
 * Enhanced Fibonacci Sequence Generator in Rust
 */

use std::io::{self, Write};

// ANSI color codes
const RESET: &str = "\x1b[0m";
const GREEN: &str = "\x1b[32m";
const YELLOW: &str = "\x1b[33m";
const CYAN: &str = "\x1b[36m";
const MAGENTA: &str = "\x1b[35m";
const RED: &str = "\x1b[31m";
const GRAY: &str = "\x1b[90m";

/// Clear the terminal screen
fn clear_screen() {
    print!("\x1b[H\x1b[2J");
    io::stdout().flush().unwrap();
}

/// Display the program header
fn display_header() {
    println!("{}==============================================={}", GREEN, RESET);
    println!("{}      Enhanced Fibonacci Sequence Generator    {}", GREEN, RESET);
    println!("{}==============================================={}", GREEN, RESET);
    println!();
}

/// Display menu options
fn display_menu() {
    println!("{}Generation Options:{}", YELLOW, RESET);
    println!("  1. Generate by number of terms");
    println!("  2. Generate up to a maximum value");
    println!("  3. Display with statistics");
    println!("  4. Display in columns");
    println!();
    print!("Enter your choice (1-4): ");
    io::stdout().flush().unwrap();
}

/// Generate and display Fibonacci sequence
fn generate_fibonacci(term_count: usize, show_stats: bool, column_mode: bool) {
    let mut t1: i64 = 0;
    let mut t2: i64 = 1;
    let mut sum: i64 = 0;
    let terms_per_line = 10;
    
    if column_mode {
        println!("{}Fibonacci Sequence ({} terms):{}", CYAN, term_count, RESET);
        println!();
    } else {
        println!("{}Fibonacci Sequence:{}", CYAN, RESET);
    }
    
    // Handle the first term
    if term_count >= 1 {
        if column_mode {
            print!("{:>12}", t1);
        } else {
            print!("{} ", t1);
        }
        sum += t1;
    }
    
    // Handle the second term
    if term_count >= 2 {
        if column_mode {
            print!("{:>12}", t2);
        } else {
            print!("{} ", t2);
        }
        sum += t2;
    }
    
    // Calculate and print remaining terms
    for j in 3..=term_count {
        let next_term = t1 + t2;
        
        if column_mode {
            print!("{:>12}", next_term);
            if j % terms_per_line == 0 {
                println!();
            }
        } else {
            print!("{} ", next_term);
        }
        
        sum += next_term;
        
        t1 = t2;
        t2 = next_term;
    }
    
    println!();
    println!();
    
    // Display statistics if requested
    if show_stats {
        println!("{}Statistics:{}", MAGENTA, RESET);
        println!("  Number of terms: {}", term_count);
        println!("  Sum of all terms: {}", sum);
        let average = sum as f64 / term_count as f64;
        println!("  Average value: {:.2}", average);
        
        if term_count >= 2 {
            println!("  Last term: {}", t2);
            println!("  Second-to-last term: {}", t1);
            
            if t1 != 0 {
                let ratio = t2 as f64 / t1 as f64;
                let golden_ratio = 1.6180339887;
                let difference = (ratio - golden_ratio).abs();
                println!("  Ratio (F(n)/F(n-1)): {:.10}", ratio);
                println!("  Golden Ratio (φ): {:.10}", golden_ratio);
                println!("  Difference: {:.10}", difference);
            }
        }
        println!();
    }
}

/// Generate Fibonacci sequence up to a maximum value
fn generate_up_to_max(max_val: i64) {
    let mut t1: i64 = 0;
    let mut t2: i64 = 1;
    let mut count = 0;
    
    println!("{}Fibonacci Sequence up to {}:{}", CYAN, max_val, RESET);
    println!();
    
    if max_val >= 0 {
        print!("{} ", t1);
        count += 1;
    }
    
    if max_val >= 1 {
        print!("{} ", t2);
        count += 1;
    }
    
    while (t1 + t2) <= max_val {
        let next_term = t1 + t2;
        print!("{} ", next_term);
        count += 1;
        t1 = t2;
        t2 = next_term;
    }
    
    println!();
    println!();
    println!("{}Total terms generated: {}{}", MAGENTA, count, RESET);
    println!();
}

/// Read an integer from standard input
fn read_int() -> Option<usize> {
    let mut input = String::new();
    io::stdin().read_line(&mut input).ok()?;
    input.trim().parse().ok()
}

/// Read an i64 from standard input
fn read_i64() -> Option<i64> {
    let mut input = String::new();
    io::stdin().read_line(&mut input).ok()?;
    input.trim().parse().ok()
}

/// Main program entry point
fn main() {
    clear_screen();
    display_header();
    display_menu();
    
    let choice = read_int().unwrap_or(0);
    println!();
    
    match choice {
        1 => {
            print!("Enter the number of terms to generate (max 92): ");
            io::stdout().flush().unwrap();
            let mut n = read_int().unwrap_or(0);
            println!();
            
            if n == 0 {
                println!("{}Error: Please enter a positive integer.{}", RED, RESET);
            } else if n > 92 {
                println!("{}Error: Maximum 92 terms supported with i64.{}", RED, RESET);
                println!("Using 92 terms instead.");
                n = 92;
                println!();
                generate_fibonacci(n, false, false);
            } else {
                generate_fibonacci(n, false, false);
            }
        }
        
        2 => {
            print!("Enter the maximum value: ");
            io::stdout().flush().unwrap();
            let max_value = read_i64().unwrap_or(-1);
            println!();
            
            if max_value < 0 {
                println!("{}Error: Please enter a non-negative value.{}", RED, RESET);
            } else {
                generate_up_to_max(max_value);
            }
        }
        
        3 => {
            print!("Enter the number of terms to generate (max 92): ");
            io::stdout().flush().unwrap();
            let mut n = read_int().unwrap_or(0);
            println!();
            
            if n == 0 {
                println!("{}Error: Please enter a positive integer.{}", RED, RESET);
            } else if n > 92 {
                println!("{}Error: Maximum 92 terms supported with i64.{}", RED, RESET);
                println!("Using 92 terms instead.");
                n = 92;
                println!();
                generate_fibonacci(n, true, false);
            } else {
                generate_fibonacci(n, true, false);
            }
        }
        
        4 => {
            print!("Enter the number of terms to generate (max 92): ");
            io::stdout().flush().unwrap();
            let mut n = read_int().unwrap_or(0);
            println!();
            
            if n == 0 {
                println!("{}Error: Please enter a positive integer.{}", RED, RESET);
            } else if n > 92 {
                println!("{}Error: Maximum 92 terms supported with i64.{}", RED, RESET);
                println!("Using 92 terms instead.");
                n = 92;
                println!();
                generate_fibonacci(n, false, true);
            } else {
                generate_fibonacci(n, false, true);
            }
        }
        
        _ => {
            println!("{}Invalid choice. Using default mode (option 1).{}", RED, RESET);
            println!();
            print!("Enter the number of terms to generate (max 92): ");
            io::stdout().flush().unwrap();
            let mut n = read_int().unwrap_or(0);
            println!();
            
            if n > 0 {
                if n > 92 { n = 92; }
                generate_fibonacci(n, false, false);
            }
        }
    }
    
    println!();
    println!("{}Press ENTER to exit...{}", GRAY, RESET);
    let mut _input = String::new();
    io::stdin().read_line(&mut _input).unwrap();
}

